### ISSUE TITLE
accents (á é í ó ú) not working under ssh if codepage set to utf8

### ISSUE THREAD

**Issue by agusmba** [git-for-windows/git#902]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen)
   or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue
   matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options
git version 2.10.0.windows.1
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver
Microsoft Windows [Versión 6.1.7601]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt
Path Option: Cmd
SSH Option: OpenSSH
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

 The problem appears when I change the terminal codepage to utf8, and only under an ssh session (in the git-bash accents work fine).

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

 bash (regular windows terminal/console)

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
# change codepage to utf8 (65001)
$ /c/Windows/System32/chcp.com 65001
Página de códigos activa: 65001

# check that accents work in bash console by typing: á é í ó ú
$ á é í ó ú (ok)

# do an ssh to a linux machine (CentOS 6 running locally in virtualbox in my case)
$ ssh root@my.server

# try to type accents a e i o u -> á é í ó ú
$ a e i o u -> 
# accented characters do not appear
```
 - What did you expect to occur after running these commands?

I expected accented characters to appear also when in an ssh session

 - What actually happened instead?

 I couldn't write accented characters

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

 not related to any repository.

**Comment by agusmba** [comment]
I upgraded to git-for-windows's new version 2.10.1, and it seems the issue is now solved (probably due to OpenSSL update?). Will do further checks when at home in order to close this if all works again.

**Comment by agusmba** [comment]
Still failing with the new version 2.10.1

**Comment by fourpastmidnight** [comment]
So, it looks like you may have deleted a comment you left earlier which I read in my email saying you believed the issue may be resolved. Now you say it isn't. So I have to ask, what lead you to think the issue may have been resolved? Did you test your scenario on another setup different from the one that is not working?

**Comment by agusmba** [comment]
Yeah sorry for the misleading comment.

The problem is that I had removed the utf8 configuration from my terminals in order to be able to work with ssh, and thus the error was not present when I tested the new version today. I had forgotten about the utf8 removal and I incorrectly thought it had been solved, but when I realized it, I re-tested it with utf8 in the terminal and the error is still there.

Again sorry for incorrectly jumping the gun

**Comment by fourpastmidnight** [comment]
No problem. Just hoped that if two environments were exhibiting different behavior it'd be pretty easy to determine what was different. 😃 

So dumb question: What's the locale setting of the remote machine?

Next question, what font is your command prompt window using?

**Comment by nalla** [comment]
Is this issue somehow `git.exe` related? Or just a unfortunate setup regarding `mintty.exe` / `ssh.exe` / `bash.exe`?

**Comment by agusmba** [comment]
No problem.

Here's the additional info:
```
[root@remote_machine ~]# locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

The terminal font is: `Lucida Console (TrueType)`

I would tend to think it is not directly related to `git.exe`, since the problem depends only on the ssh client when the terminal is in utc8 mode. There is no link that I know about to `git.exe`.

Note that I'm not using the mintty install but the regular-windows-console one.

**Comment by nalla** [comment]
Where is your `ssh.exe` comming from? how do you start your `bash.exe` if your're not using `mintty`? I'dont think this is a git for windows bug.

**Comment by agusmba** [comment]
Hi again,

What additional info is missing? I'm willing to test/provide any info I can reasonably get.

BTW, I'd rather not enter into mintty territory since I'm not using it (and it seems it may have its own problems related to accents), I propose to focus on the windows terminal for this issue.

**Comment by nalla** [comment]
> What additional info is missing?

Please read my previous commit.

> Where is your ssh.exe coming from? How do you start your bash.exe if your're not using mintty?

**Comment by agusmba** [comment]
> Where is your ssh.exe comming from?

It's the one provided by git-for-windows:

```
USER@LOCALHOST MINGW64 ~
$ which ssh
/usr/bin/ssh
$ ssh -V
OpenSSH_7.3p1, OpenSSL 1.0.2j  26 Sep 2016
```

> how do you start your bash.exe if your're not using mintty? I'dont think this is a git for windows bug.

I'm using the link installed by git-for-windows: It seems to point to `"C:\Program Files\Git\git-bash.exe" --cd-to-home`

Only when I installed git-for-windows I use the option to use window's own terminal instead of mintty, but It's still a normal git-for-windows installation.

**Comment by agusmba** [comment]
I don't have vagrant installed, but if you think it would help others reproduce the problem, I'd be willing to install it in order to share a simple vagrant command to run a linux machine (instead of doing it manually with virtualbox as I'm doing) and modify the steps-to-reproduce so that someone else can confirm this.

**Comment by nalla** [comment]
So you are relying on the fact that a call to `/c/Windows/System32/chcp.com 65001` will be honored by the

```
$ help
GNU bash, version 4.3.46(2)-release (x86_64-pc-msys)
```

msys2 bash application that is then calling another msys2 application (ssh.exe)... I'm out here. That's beyond my expertise. But my opinion stands, this is not a _Git for Windows_ issue. You might find better help in the msys2 community. Not here.

**Comment by agusmba** [comment]
> msys2 bash application that is then calling another msys2 application (ssh.exe)... I'm out here. That's beyond my expertise. But my opinion stands, this is not a Git for Windows issue. You might find better help in the msys2 community. Not here.

Ok, fair enough. I initially opened the but against ConEmu since I was experiencing it there, then with @Maximus5 help narrowed it down to git-for-windows installation, but you are also right in the sense that it has nothing to do with `git` per se. The problem seems to be in one of the bundled "extras".

I'll open yet another bug report there. But since I have no idea about Msys2, I will be at a loss when they ask me to replicate the bug using only their tools. Let's hope not.

Thanks for your time!

**Comment by Maximus5** [comment]
That's the usual problem of all ported tools. Nobody is responsible for it.
SSH, Vim or dozens of other tools were initially developed for Linux by certain authors.  And you are here

* sometimes authors abandoned their product for community support
* authors do not support Windows builds because they don't use Windows
* cygwin/msys maintainers do "rebuild" of these tools with their ported "cores"

**Comment by agusmba** [comment]
Yeah, I understand. I'd like to thank you @Maximus5, @nalla and all the git-for-windows team, and the msys2 guys for making such great tools.

I'd move to linux directly if I could but your tools make working in windows a lot more bearable!

**Comment by dscho** [comment]
@agusmba I verified that this issue persists with MSYS2: if you start a `cmd.exe`, `chcp 65001` and then execute `<MSYS2ROOT>\usr\bin\ssh.exe <HOST>`, you cannot type accented characters, either.

However, MSYS2 really suggests very strongly to start in a MinTTY, and the accented characters *do* work there.

On the other hand, I just tested Cygwin and it exposes the very same problem, and Cygwin *does* run in a Windows console by default.

So if I were you, I would try to solicit advice from the Cygwin project (look for IRC or for the mailing list on the left side of https://cygwin.com/) how to investigate this further. It would be nice if you would loop back with any further developments on this issue (or when you get stuck).
